### PR TITLE
feat(datasource): convert to ts

### PR DESCRIPTION
**Pull request by viceice**
This pr will convert the dfatasources to typescript.

depends on: #4180, #4231

### CODE REVIEW

**Comment by ViceIce** [comment]
@rarkins Here we can convert datasource by datasource, so i can split to multiple pr if you like.

File: lib/datasource/ruby-version/index.ts [hunk 1]
```diff
@@ -1,25 +1,27 @@
...
 const rubyVersionsUrl = 'https://www.ruby-lang.org/en/downloads/releases/';
 
-async function getPkgReleases() {
+export async function getPkgReleases(
+  _config?: PkgReleaseConfig
+): Promise<ReleaseResult> {
   // First check the persistent cache
   const cacheNamespace = 'datasource-ruby-version';
-  const cachedResult = await renovateCache.get(cacheNamespace, 'all');
+  const cachedResult = await renovateCache.get<ReleaseResult>(
```
Comment (rarkins) [inline lib/datasource/ruby-version/index.ts]: This is not `Promise<ReleaseResult>`, or doesn't matter eitherway?

Comment (ViceIce) [reply lib/datasource/ruby-version/index.ts]: seems ok for ts / js

**Comment by rarkins** [comment]
😨

**Comment by ViceIce** [comment]
❓

**Comment by rarkins** [comment]
Time to smoke test :)

**Pull request merged**